x-pack/plugins/fleet/server/services/secrets.ts

```diff
@@ -0,0 +1,296 @@
+        outcome: 'success',
+      },
+    });
+  } catch (e) {
+    const logger = appContextService.getLogger();
+    const msg = `Error deleting secret '${id}' from ${getSecretsIndex()} index: ${e}`;
+    logger.error(msg);
+    throw new FleetError(msg);
+  }
+
+  return res.result;
+}
+
+export async function extractAndWriteSecrets(opts: {
+  packagePolicy: NewPackagePolicy;
+  packageInfo: PackageInfo;
+  esClient: ElasticsearchClient;
+}): Promise<{ packagePolicy: NewPackagePolicy; secret_references: PolicySecretReference[] }> {
+  const { packagePolicy, packageInfo, esClient } = opts;
+  const secretPaths = getPolicySecretPaths(packagePolicy, packageInfo);
+
+  if (!secretPaths.length) {
+    return { packagePolicy, secret_references: [] };
+  }
+
+  const secrets = await createSecrets({
+    esClient,
+    values: secretPaths.map((secretPath) => secretPath.value.value),
+  });
+
+  const policyWithSecretRefs = JSON.parse(JSON.stringify(packagePolicy));
+  secretPaths.forEach((secretPath, i) => {
+    set(policyWithSecretRefs, secretPath.path + '.value', toVarSecretRef(secrets[i].id));
+  });
+
+  return {
+    packagePolicy: policyWithSecretRefs,
+    secret_references: secrets.map(({ id }) => ({ id })),
+  };
+}
+
+function isSecretVar(varDef: RegistryVarsEntry) {
+  return varDef.secret === true;
+}
+
+function containsSecretVar(vars?: RegistryVarsEntry[]) {
+  return vars?.some(isSecretVar);
+}
+
+// this is how secrets are stored on the package policy
+function toVarSecretRef(id: string): VarSecretReference {
+  return { id, isSecretRef: true };
+}
+
+// this is how IDs are inserted into compiled templates
+export function toCompiledSecretRef(id: string) {
+  return `$co.elastic.secret{${id}}`;
+}
+
+// Given a package policy and a package,
+// returns an array of lodash style paths to all secrets and their current values
+export function getPolicySecretPaths(
+  packagePolicy: PackagePolicy | NewPackagePolicy,
+  packageInfo: PackageInfo
+): SecretPath[] {
+  const packageLevelVarPaths = _getPackageLevelSecretPaths(packagePolicy, packageInfo);
+
+  if (!packageInfo?.policy_templates?.length || packageHasNoPolicyTemplates(packageInfo)) {
+    return packageLevelVarPaths;
+  }
+
+  const inputSecretPaths = _getInputSecretPaths(packagePolicy, packageInfo);
+
+  return [...packageLevelVarPaths, ...inputSecretPaths];
+}
+
+function _getPackageLevelSecretPaths(
+  packagePolicy: NewPackagePolicy,
+  packageInfo: PackageInfo
+): SecretPath[] {
+  const packageSecretVars = packageInfo.vars?.filter(isSecretVar) || [];
+  const packageSecretVarsByName = keyBy(packageSecretVars, 'name');
+  const packageVars = Object.entries(packagePolicy.vars || {});
+
+  return packageVars.reduce((vars, [name, configEntry], i) => {
+    if (packageSecretVarsByName[name]) {
+      vars.push({
+        value: configEntry,
+        path: `vars.${name}`,
+      });
+    }
+    return vars;
+  }, [] as SecretPath[]);
+}
+
+function _getInputSecretPaths(
```

Comment:
this function is a bit too complex, could be broken down to smaller functions

Reply:
split some logic out in [b74d0cf](https://github.com/elastic/kibana/pull/156036/commits/b74d0cfceda24db2f663fdaa23cdac1458b24753)